spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseFeatures.groovy

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2010 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.spockframework.smoke.extension
+
+import org.spockframework.EmbeddedSpecification
+import org.spockframework.runtime.InvalidSpecException
+
+class StepwiseFeatures extends EmbeddedSpecification {
+  def "basic usage"() {
+    runner.throwFailure = false
+
+    when:
+    def result = runner.runWithImports("""
+class Foo extends Specification {
+  @Stepwise
+  def step1() {
+    expect: count != 3
+    where: count << (1..5)
+  }
+}
+    """)
+    def expectedSkipMessagesCount = result.testEvents()
+      .filter({ event ->
+        event.payload.present &&
+          event.payload.get() == "skipping subsequent iterations after failure"
+      })
+      .count()
+
+    then:
+    result.testsStartedCount == 3 + 1
+    result.testsSucceededCount == 2 + 1
+    result.testsFailedCount == 1
+    result.testsSkippedCount == 2
+    expectedSkipMessagesCount == 2
+  }
+
+  def "basic usage rolled-up"() {
+    runner.throwFailure = false
+    runner.addClassImport(RollupIterationCounter)
+    RollupIterationCounter.count.set(0)
+
+    when:
+    def result = runner.runWithImports("""
+class Foo extends Specification {
+  @Stepwise
+  @Rollup
+  def step1() {
+    // Keep track of the number of started iterations for later verification in the outer feature method
+    RollupIterationCounter.count.set(RollupIterationCounter.count.get() + 1)
+    expect: count != 3
+    where: count << (1..5)
+  }
+}
+    """)
+
+    then:
+    RollupIterationCounter.count.get() == 3
+    result.testsStartedCount == 1
+    result.testsSucceededCount == 0
+    result.testsFailedCount == 1
+    result.testsSkippedCount == 0
+  }
+
+  /**
+   * Has the sole purpose of reporting back the number of started iterations from the rolled-up, stepwise feature method
+   * used in the embedded spec under test in feature "basic usage rolled-up".
+   *
+   * TODO: find a better way to communicate with the embedded spec
+   */
+  private static class RollupIterationCounter {
+    static InheritableThreadLocal<Integer> count = new InheritableThreadLocal<>()
+  }
```

Comment:
Will you resolve that?
In `org.spockframework.smoke.CleanupBlocks` a similar pattern is used, although you don't need an inner class for that.

Reply:
I asked for your feedback concerning this issue in Slack, because I had no better idea. Looking at `CleanupBlocks`, you are using a static member, which is used in two methods. That would not be thread-safe, so you mitigate it using `@Execution(SAME_THREAD)`. I am not sure if that is better. I explicitly wanted my solution to work when running tests concurrently, so I chose a thread-local field. Because I also wanted to avoid the embedded spec to import the complete surrounding spec, I isolated it in an inner static class.

So basically, we are using the same type of solution: We communicate results from the embedded spec to the surrounding class using some kind of static member. That answers my question from Slack if there is a better way, e.g. binding an variable in the embedded spec directly. I guess, this is impossible at present, otherwise you would have done it in `CleanupBlocks`. So my choice would be to not change anything there for now, because my variant should work without `@Execution(SAME_THREAD)`. Would that be OK? If you insist, I can remove the TODO, but actually I like to keep TODOs as reminders for the future and as incentive for co-developers to think about it and come up with a better solution, if the architecture of Spock ever changes in a way enabling such communication more easily, just like with the test results which are communicated to the outside of the spec.

Reply:
For the record after talking privately: We keep that static class with the thread-local (not inherited, though, because according to @leonard84 embedded specs run in the same thread as the feature starting them. The TODO goes away.